CMakeLists.txt

```diff
@@ -66,13 +72,25 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${FLAGS_COMMON} -D__STDC_LIMIT_MACROS -D
 set(RR_FLAGS_DEBUG "-Wall -Wextra -DDEBUG -UNDEBUG")
 set(RR_FLAGS_RELEASE "-Wall -Wextra -UDEBUG -DNDEBUG")
 
+# When building RPMs, we need to enable the following options in order to avoid errors
+# arising from -D_FORTIFY_SOURCE=2 and -Werror. We check the availability of the options
+# and use them accordingly.
+CHECK_CXX_COMPILER_FLAG("-Wno-error=unused-result" SUPPORTS_NO_ERROR_UNUSED_RESULT)
+CHECK_CXX_COMPILER_FLAG("-Wno-error=attribute-warning" SUPPORTS_NO_ERROR_ATTRIBUTE_WARNING)
+if (SUPPORTS_NO_ERROR_UNUSED_RESULT)
+  set(CXX_NO_ERROR "-Wno-error=unused-result")
+endif()
+if (SUPPORTS_NO_ERROR_ATTRIBUTE_WARNING)
+  set(CXX_NO_ERROR "${CXX_NO_ERROR} -Wno-error=attribute-warning")
+endif()
+
 # The folowing settings are the defaults for the OTHER build type.
 # Flags used to build the preload library. MUST have debuginfo enabled. SHOULD be optimized.
 set(PRELOAD_COMPILE_FLAGS "${RR_FLAGS_RELEASE} -fno-stack-protector -g3")
 # Flags used to build Brotli. SHOULD be optimized. MUST NOT error on warnings.
 set(BROTLI_COMPILE_FLAGS ${RR_FLAGS_RELEASE})
 # Flags used to build tests. MUST have -DDEBUG and debuginfo enabled, MUST NOT be optimized.
-set(RR_TEST_FLAGS "${RR_FLAGS_DEBUG} -g3 -O0")
+set(RR_TEST_FLAGS "${RR_FLAGS_DEBUG} -g3 -Og")
```

Comment:
What's the purpose of this change?

Reply:
Using -O0 causes the builds to fail because of the additional fortify checks that are added when we build the RPM (_FORTIFY_SOURCE requires compiling with optimization). Using -Og solves this problem and hopefully doesn't introduce too much optimization.

Reply:
Do tests actually pass with `-Og`? Some tests are sensitive to the optimization level.

Can't you pass `-N_FORTIFY_SOURCE` or something like that instead to disable that explicitly?

Reply:
On recent run on RHEL7 all but 7 tests passed using the -Og. On Fedora IIRC it was 3 test failures, so things appear to work with the -Og.  Is there somewhere that describes the environment and which of the tests has issues that resulted in -O0 being used?

The reason for resetting the _FORTIFY_SOURCE in the only read_oversize.c was to keep as much of the checking on as possible for the code base.  Turning it off for the entire build would be undesirable. Similar strategy for leaving the -Werror and carving out the warning that appear harmless in the testsuite with -Wno-error=... .

Reply:
`RR_TEST_FLAGS` only affects the tests, and I think disabling `_FORTIFY_SOURCE` for the tests is a good thing. Many of the tests intentionally do all sorts of weird and potentially invalid things; those tests are not intended to be "valid" C programs. (Because of course we need to test that rr reliably handles various kinds of invalid C programs.)

> Is there somewhere that describes the environment and which of the tests has issues that resulted in -O0 being used?

No. The main issue that springs to mind is that many tests rely on there being no inlining.

Reply:
And yes, I know `-Og` typically disables inlining, but that isn't guaranteed.

Reply:
From gcc manual (https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html):

-Og

    Optimize debugging experience. -Og should be the optimization level of choice for the standard edit-compile-debug cycle, offering a reasonable level of optimization while maintaining fast compilation and a good debugging experience. It is a better choice than -O0 for producing debuggable code because some compiler passes that collect debug information are disabled at -O0.

    Like -O0, -Og completely disables a number of optimization passes so that individual options controlling them have no effect. Otherwise -Og enables all -O1 optimization flags except for those that may interfere with debugging:

    -fbranch-count-reg  -fdelayed-branch 
    -fdse  -fif-conversion  -fif-conversion2  
    -finline-functions-called-once 
    -fmove-loop-invariants  -fssa-phiopt 
    -ftree-bit-ccp  -ftree-dse  -ftree-pta  -ftree-sra

Clang also lists "-Og" (https://clang.llvm.org/docs/CommandGuide/clang.html). However, doesn't list which of the -O1 optimizations are disabled. Given the description it could be possible for some problematic optimizations to occur with the clang compiler.

Reply:
@wcohen I think that the recent tests that I did on RHEL7 did not have `-Og` set. However, we did run the tests before that which did have `-Og` set and most tests were passing.

I'm running the testsuite again to check for this.

Reply:
After rerunning the testsuite with and without the above configuration, it seems that `-Og` does break a lot of the test cases in RHEL7. This seems to be the case for Fedora as well (testsuite still running). When I had ran the testsuite previously, I **didn't** actually have the `-Og` option set. 

I'll try to find a workaround.